### PR TITLE
Pandas DataFrames for tally data

### PR DESCRIPTION
**Pull request by wbinventor**
This PR introduces a number of new features to the Python API to facilitate tally data processing and maniuplation. Many of the new features introduced in this PR can be viewed on this public [IPython Notebook](http://nbviewer.ipython.org/gist/wbinventor/13b4d9a15fb0de0dc247) which illustrates how to create a model, run an OpenMC simulation, and perform tally data processing within the interactive IPython environment.

First and foremost, a new ``Tally.get_pandas_dataframe(...)`` routine is introduced to address issue #382. [Pandas](http://pandas.pydata.org/) enables statistical analysis and visualization of complicated datasets. I hope that the power of Pandas to help a user view and manipulate ``Tally`` data is well illustrated in the IPython Notebook.

In addition, the ``StatePoint.get_tally(...)`` is modified to make it easier to use - it no longer requires a user to pass in ``Filters``, ``Nuclides``, scores, etc. which define the complete specification for the ``Tally`` to extract. Only a subset of the specifications need to be passed into the routine now which can make it much easier to extract ``Tallies`` from the ``StatePoint`` API (in particular, simply specifying the "name" of the ``Tally`` may often be enough).  

The ``Tally.get_values(...)`` routine is also completely refactored to make it easier to extract ``Tally`` data. This routine now may return a scalar *or* vector-valued data based on the users input parameters. This is illustrated in the IPython Notebook.

I should also note that a few bugs in "statepoint.py" and "tallies.py" related to the striding of tally data are corrected in this PR.

Finally, This PR builds upon #379 (distribcell tallies) and would be most easily reviewed after that PR is merged into the code.

### CODE REVIEW

**Comment by wbinventor** [comment]
I have no idea why Travis is failing so spectacularly for this since all of the tests work fine on my machine.

**Comment by wbinventor** [comment]
This is now passing all tests except "test_filter_distribcell". Unfortunately, I cannot reproduce this failure on my machine.

**Comment by paulromano** [comment]
Here's the traceback:
```
24: ../../src/utils/openmc/statepoint.py:573: RuntimeWarning: invalid value encountered in double_scalars
24:   self._global_tallies[i, 1] = t_value * np.sqrt((s2 / n - s**2) / (n-1))
24: ../../src/utils/openmc/tallies.py:408: RuntimeWarning: invalid value encountered in divide
24:   self.mean**2) / (self.num_realizations - 1))
24: Traceback (most recent call last):
24:   File "results.py", line 93, in <module>
24:     outstr += "{0:12.6E}\n".format(sp3_t4_d[()])
24: ValueError: Unknown format code 'E' for object of type 'str'
24: Traceback (most recent call last):
24:   File "test_filter_distribcell.py", line 106, in <module>
24:     test_results()
24:   File "test_filter_distribcell.py", line 83, in test_results
24:     compare = filecmp.cmp('results_test.dat', 'results_true.dat')
24:   File "/usr/lib/python2.7/filecmp.py", line 42, in cmp
24:     s1 = _sig(os.stat(f1))
24: OSError: [Errno 2] No such file or directory: 'results_test.dat'
````

I figured out a way to get more output on Travis. If you change the -V argument on line 166 of run_tests.py to -VV, you'll see the actual output from each test run

**Comment by wbinventor** [comment]
Awesome, thanks for that traceback and the hint for debugging Travis runs. A very minor update was needed for "test_filter_distribcell" and now it passes along with the other tests.

File: src/utils/openmc/statepoint.py [hunk 1]
```diff
@@ -652,41 +668,31 @@ def get_tally(self, score, filters, nuclides,
...
 
 
-    def get_tally_id(self, score, filters, name='', estimator='tracklength'):
-        """Retrieve the Tally ID for a given list of filters and score(s).
+    def link_with_summary(self, summary):
+        """Links Tallies and Filters with Summary model information.
+
+        This routine retrieves model information (materials, geometry) from a
+        Summary object populated with an HDF5 'summary.h5' file and inserts
+        it into the Tally objects. This can be helpful when ciewing and
```
Comment (paulromano) [inline src/utils/openmc/statepoint.py]: typo "ciewing"

File: src/utils/openmc/tallies.py [hunk 2]
```diff
@@ -532,161 +545,560 @@ def get_tally_xml(self):
...
+                        try:
+                            import opencg
+                        except ImportError:
+                            msg = 'The OpenCG package must be installed ' \
+                                  'to use a Summary for distribcell dataframes'
+                            raise ImportError(msg)
+
+                        # Create and extract the OpenCG geometry the Summary
+                        summary.make_opencg_geometry()
+                        opencg_goemetry = summary.opencg_geometry
```
Comment (paulromano) [inline src/utils/openmc/tallies.py]: typo "goemetry". Three more instances of this below as well

File: src/utils/openmc/tallies.py [hunk 3]
```diff
@@ -532,161 +545,560 @@ def get_tally_xml(self):
...
+                                # Move to next node in LocalCoords linked list
+                                if coords._next == None:
+                                    offsets_to_coords[offset] = None
+                                else:
+                                    offsets_to_coords[offset] = coords._next
+                                    levels_remain = True
+
+                            # Tile the Multi-index columns
+                            for level_key, level_bins in level_dict.items():
+                                level_bins = np.repeat(level_bins,filter.stride)
```
Comment (paulromano) [inline src/utils/openmc/tallies.py]: space between arguments

File: src/utils/openmc/tallies.py [hunk 4]
```diff
@@ -532,161 +545,560 @@ def get_tally_xml(self):
...
+                        # Each distribcell offset is a DataFrame bin
+                        # Unravel the paths into DataFrame columns
+                        num_offsets = len(offsets_to_coords)
+
+                        # Initialize termination condition for while loop
+                        levels_remain = True
+                        counter = 0
+
+                        # Iterate over each level in the CSG tree hierarchy
+                        while(levels_remain):
```
Comment (paulromano) [inline src/utils/openmc/tallies.py]: `while levels_remain:`

**Comment by paulromano** [comment]
Wow -- this feature is awesome! Two small requests to prevent users (like myself) from running into issues:

1. I tried to use the `Tally.get_values` method with having called `StatePoint.read_results` and it just crashed. Can you put in a check to prevent this or give a sensible error message?
2. I tried to use `Tally.get_pandas_dataframe(summary=...)` without having called `StatePoint.link_with_summary()`. Again, I got a not very useful exception. Same request here: a sensible error message informing the user they need to link with a summary first.

**Comment by wbinventor** [comment]
All of your comments have been addressed @paulromano

**Comment by wbinventor** [comment]
I just added an optional ``id`` parameter to the ``StatePoint.get_tally(...)`` routine (see 310d2a8). As @bforget has pointed out to me, it can be very useful and intuitive to simply access tallies in the statepoint files when a user only hard-codes a few tallies (and their IDs) in ``tallies.xml``. The new method signature is as follows:

```python
StatePoint.get_tally(scores=[], filters=[], nuclides=[], name=None, id=None, estimator=None)
```

I think this now provides the easy-to-use access ability via the ``id`` and/or ``name`` parameters, as well as the more detailed parameters needed for power users.

**Pull request merged**